### PR TITLE
Provides an async download API, a DownloadManager and a ZipBuilder for the storage framework.

### PR DESCRIPTION
**Pull request by andyHa**
ObjectStores and the ObjectStorageSpaces now support an asynchronous API to fetch objects.
Additionaly the DownloadManager can be used to download multiple files in parallel (if possible).
Finally the ZipBuilder can be used to create a ZIP archive (or more if an upper limit is reached).

### CODE REVIEW

**Review by qw3ry — APPROVED** [review]

File: src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java [hunk 1]
```diff
@@ -323,6 +366,15 @@ protected abstract void storePhysicalObject(String objectKey, InputStream data,
     @Nullable
     protected abstract FileHandle getData(String objectKey) throws IOException;
 
+    /**
+     * Downloads an provides the contents of the requested object asynchronous.
```
Comment (mkeckmkeck) [inline src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java:370]:
```suggestion
     * Downloads and provides the contents of the requested object asynchronous.
```

File: src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java [hunk 2]
```diff
@@ -334,6 +386,16 @@ protected abstract void storePhysicalObject(String objectKey, InputStream data,
     @Nullable
     protected abstract FileHandle getData(String objectKey, ByteBlockTransformer transformer) throws IOException;
 
+    /**
+     * Downloads an provides the contents of the requested object asynchronous.
```
Comment (mkeckmkeck) [inline src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java:390]:
```suggestion
     * Downloads and provides the contents of the requested object asynchronous.
```

File: src/main/java/sirius/biz/storage/layer1/DownloadManager.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Permits to download several objects from a {@link ObjectStorageSpace}.
+ * <p>
+ * If possible (and feasible) the downloads will be executed asynchronously to increase the overall speed. However,
+ * the callbacks to the completion consmer will always happen from the thread which invokes {@link #close()} or
```
Comment (mkeckmkeck) [inline src/main/java/sirius/biz/storage/layer1/DownloadManager.java:28]:
```suggestion
 * the callbacks to the completion consumer will always happen from the thread which invokes {@link #close()} or
```

**Review by mkeckmkeck — APPROVED** [review]
typos

**Review by fhaScireum — APPROVED** [review]

**Review by cschierle — APPROVED** [review]

**Pull request merged**